src/node/blockstorage.cpp

```diff
@@ -1019,21 +1022,18 @@ bool BlockManager::ReadBlock(CBlock& block, const FlatFilePos& pos) const
         return false;
     }
 
+    if (expected_hash && block_hash != *expected_hash) {
+        LogError("GetHash() doesn't match index at %s while reading block", pos.ToString());
+        return false;
+    }
+
     return true;
 }
 
 bool BlockManager::ReadBlock(CBlock& block, const CBlockIndex& index) const
 {
     const FlatFilePos block_pos{WITH_LOCK(cs_main, return index.GetBlockPos())};
-
-    if (!ReadBlock(block, block_pos)) {
-        return false;
-    }
-    if (block.GetHash() != index.GetBlockHash()) {
```

Comment:
Is there really a scenario where we could be reading a block, but not have its block index entry?  Maybe this should just be an `Assume` instead.

Reply:
We're not always checking this, only when the index is available.
But after this change this check is for free, no need to recalculate the header's hash anymore.

Reply:
> I can't see any

Most usages have a hash available, but I'm not sure about e.g. [`LoadExternalBlockFile`](https://github.com/bitcoin/bitcoin/blob/master/src/validation.cpp#L5166) - but even if I would, I'd prefer tackling it separately, I'm not comfortable adding new exceptions to consensus code.

Reply:
re: https://github.com/bitcoin/bitcoin/pull/32487#discussion_r2094072165

I don't think I understand suggestion to use Assume. But if the hash is always available I guess expected_hash could be a required parameter instead of an optional one. Would seem fine to expand the PR, or just keep it focused on not calculating the hash twice.